### PR TITLE
feat[contracts, l2geth]: backend mechanics for OVM self-upgrades

### PR DESCRIPTION
**Pull request by ben-chain**
**Description**
This PR to the 0.4.0 release candidate gives the execution manager access to directly setting account code and storage, so that versioning can be performed from within the L2 state transitions.

This is a rebase/rename of https://github.com/ethereum-optimism/optimism/pull/682 onto `v0.4.0-rc`; I have addressed all PR review which was given on that PR.

### CODE REVIEW

File: integration-tests/test/ovm-self-upgrades.spec.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      l2Wallet
+    )
+
+    Factory__ReturnOne = await ethers.getContractFactory('ReturnOne', l2Wallet)
+    const Factory__ReturnTwo = await ethers.getContractFactory(
+      'ReturnTwo',
+      l2Wallet
+    )
+    const returnTwo = await (
+      await Factory__ReturnTwo.deploy()
```
Comment (tynes) [inline integration-tests/test/ovm-self-upgrades.spec.ts:104]: Nested awaits are an anti pattern imo

File: integration-tests/test/ovm-self-upgrades.spec.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+          target: DummyContract.address,
+          code:
+            '0x1234123412341234123412341234123412341234123412341234123412341234',
+        },
+      ]
+      await applyAndVerifyUpgrade(basicCodeUpgrade)
+    })
+
+    it('Should execute a basic code upgrade which is not overwriting an existing account', async () => {
+      // TODO: fix me?  Currently breaks due to nil pointer dereference; triggerd by evm.StateDB.SetCode(...) in ovm_state_manager.go ?
```
Comment (tynes) [inline integration-tests/test/ovm-self-upgrades.spec.ts:147]: Could you give an update on this comment?

**Comment by tynes** [comment]
Do you mind adding a changeset?

**Comment by tynes** [comment]
What is the plan with the `OVM_UpgradeExecutor`? This PR has a temp implementation

File: integration-tests/contracts/SimpleStorage.sol [hunk 3]
```diff
@@ -5,9 +5,9 @@ contract ICrossDomainMessenger {
 }
 
 contract SimpleStorage {
+    bytes32 public value;
     address public msgSender;
     address public xDomainSender;
-    bytes32 public value;
```
Comment (smartcontracts) [inline integration-tests/contracts/SimpleStorage.sol:10]: Why did you move this variable? Just curious

File: integration-tests/test/ovm-self-upgrades.spec.ts [hunk 4]
```diff
@@ -0,0 +1,202 @@
+import { expect } from 'chai'
+import { Wallet, utils, BigNumber, Contract, ContractFactory } from 'ethers'
+
+import { OptimismEnv } from './shared/env'
+
+import { getContractInterface } from '@eth-optimism/contracts'
+import { l2Provider, OVM_ETH_ADDRESS } from './shared/utils'
+import { ethers } from 'hardhat'
+
+// TODO: use actual imported Chugsplash type
```
Comment (smartcontracts) [inline integration-tests/test/ovm-self-upgrades.spec.ts:10]: You should be able to get these types from `@eth-optimism/conracts` now (`ChugSplashAction`).

File: integration-tests/test/ovm-self-upgrades.spec.ts [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+        },
+      ]
+
+      await applyAndVerifyUpgrade(storageVarUpgrade)
+
+      const valueAfter = await SimpleStorage.value()
+      expect(valueAfter).to.eq(newValue)
+    })
+
+    it('code with an updated constant returns the new constant', async () => {
```
Comment (smartcontracts) [inline integration-tests/test/ovm-self-upgrades.spec.ts:183]: I'd like to see a few more aggressive tests that change mission-critical code, e.g., change `OVM_ExecutionManager` itself.

Comment (smartcontracts) [reply integration-tests/test/ovm-self-upgrades.spec.ts:183]: Here's a list I can think of:
1. Modify `OVM_ExecutionManager`.
2. Modify storage in `OVM_ExecutionManager`.
3. Modify the `OVM_SequencerEntrypoint`, `OVM_ProxyEOA`, and `OVM_ECDSAContractAccount`.
4. Modify `OVM_ETH`.

**Review by smartcontracts — CHANGES_REQUESTED** [review]
Looking good. Requesting a few more tests be added to handle more potentially breaking contracts.

File: integration-tests/contracts/SimpleStorage.sol (hunk 3 above)
Comment (ben-chain) [reply integration-tests/contracts/SimpleStorage.sol:10]: First variable is the easiest key in the whole storage layout ;)

File: integration-tests/test/ovm-self-upgrades.spec.ts (hunk 5 above)
Comment (ben-chain) [reply integration-tests/test/ovm-self-upgrades.spec.ts:183]: WRT 2. here: I actually think that modifying storage of the EM is actually something that is off-limits.  Because it isn't virtualized within the `OVM_StateManager.contractStorage` mapping, it's gonna be a pain to get that into L1 correctly.  I think that the correct model (also relevant for L1 refunds if storing `0x00...` and minimizing cost during the fraud proof) we should follow is that EM storage slots are always reset (i.e. `EM._resetContext`) at the end of execution.  This also means we should consider setting the first 50 slots to a nonzero value in case we want more variables in the future.

File: integration-tests/test/ovm-self-upgrades.spec.ts (hunk 2 above)
Comment (ben-chain) [reply integration-tests/test/ovm-self-upgrades.spec.ts:147]: Removing the comment--I have not been able to repro since the first couple days I was implementing this.  Extensive testing is definitely needed :)

**Comment by snario** [comment]
This needs to re-target `regenesis/0.4.0`.

**Comment by smartcontracts** [comment]
Gonna take this PR over from here as per @ben-chain's request

**Comment by gakonst** [comment]
Rebased on latest `regenesis/0.4.0`, no further changes made.

**Comment by ben-chain** [comment]
this has been incorporated into other branches, closing.
